src/hotspot/share/runtime/arguments.cpp

```diff
@@ -585,26 +585,33 @@ static AliasedFlag const aliased_jvm_flags[] = {
-  { "PrintSharedSpaces",         LogLevel::Info,  true,  LOG_TAGS(cds) },
-  { "TraceBiasedLocking",        LogLevel::Info,  true,  LOG_TAGS(biasedlocking) },
-  { "TraceClassLoading",         LogLevel::Info,  true,  LOG_TAGS(class, load) },
-  { "TraceClassLoadingPreorder", LogLevel::Debug, true,  LOG_TAGS(class, preorder) },
-  { "TraceClassPaths",           LogLevel::Info,  true,  LOG_TAGS(class, path) },
-  { "TraceClassResolution",      LogLevel::Debug, true,  LOG_TAGS(class, resolve) },
-  { "TraceClassUnloading",       LogLevel::Info,  true,  LOG_TAGS(class, unload) },
-  { "TraceExceptions",           LogLevel::Info,  true,  LOG_TAGS(exceptions) },
-  { "TraceInvokeDynamic",        LogLevel::Debug, true,  LOG_TAGS(methodhandles, indy) },
-  { "TraceLoaderConstraints",    LogLevel::Info,  true,  LOG_TAGS(class, loader, constraints) },
-  { "TraceMethodHandles",        LogLevel::Info,  true,  LOG_TAGS(methodhandles) },
-  { "TraceMonitorInflation",     LogLevel::Trace, true,  LOG_TAGS(monitorinflation) },
-  { "TraceSafepointCleanupTime", LogLevel::Info,  true,  LOG_TAGS(safepoint, cleanup) },
-  { "TraceJVMTIObjectTagging",   LogLevel::Debug, true,  LOG_TAGS(jvmti, objecttagging) },
-  { "TraceRedefineClasses",      LogLevel::Info,  false, LOG_TAGS(redefine, class) },
-  { "PrintJNIResolving",         LogLevel::Debug, true,  LOG_TAGS(jni, resolve) },
   { NULL,                        LogLevel::Off,   false, LOG_TAGS(_NO_TAG) }
 };
 
+// Use this to obsolete old tracing flags and suggest logging alternative.
+static AliasedObsoleteLoggingFlag const removed_product_logging_flags[] = {
+  { "PrintSharedSpaces",         "-Xlog:cds=",                 "info",  "16.0" },
+  { "TraceBiasedLocking",        "-Xlog:biasedlocking=",       "info",  "16.0" },
+  { "TraceClassLoading",         "-Xlog:class+load=",          "info",  "16.0" },
+  { "TraceClassLoadingPreorder", "-Xlog:class+preorder=",      "debug", "16.0" },
+  { "TraceClassPaths",           "-Xlog:class+path=",          "info",  "16.0" },
+  { "TraceClassResolution",      "-Xlog:class+resolve=",       "debug", "16.0" },
+  { "TraceClassUnloading",       "-Xlog:class+unload=",        "info",  "16.0" },
+  { "TraceExceptions",           "-Xlog:exceptions=",          "info",  "16.0" },
+  { "TraceInvokeDynamic",        "-Xlog:methodhandles+indy=",  "debug", "16.0" },
+  { "TraceLoaderConstraints",    "-Xlog:loader+constraints=",  "info",  "16.0" },
+  { "TraceMethodHandles",        "-Xlog:methodhandles=",       "info",  "16.0" },
+  { "TraceMonitorInflation",     "-Xlog:monitorinflation=",    "trace", "16.0" },
+  { "TraceSafepointCleanupTime", "-Xlog:safepoint+cleanup=",   "info",  "16.0" },
+  { "TraceJVMTIObjectTagging",   "-Xlog:jvmti+objecttagging=", "debug", "16.0" },
+  { "TraceRedefineClasses",      "-Xlog:redefine+class=",      "info",  "16.0" },
+  { "PrintJNIResolving",         "-Xlog:jni+resolve=",         "debug", "16.0" },
+  { NULL, NULL, NULL, NULL }
```

Comment:
I think if we wanted to give a message that the flag was obsolete and to suggest changing the command line, we should only do it for -XX:+TraceClassLoading and -XX:+TraceExceptions (I'd originally thought -XX:+TraceClassUnloading was important enough to release note but now I'm not so sure.)   The rest of the flags should either go in the table that they're no longer recognized.

Reply:
Keeping the message for any flag requires keeping all the supporting code. I don't see the "big 3" are special. They have been deprecated since 9 and we have clearly told people this when they use them. We're also release-noting this for 16 (again - this was documented when UL was added). I don't think we have to pander to anyone who hasn't updated their launch scripts by now.

Reply:
For these three, I kind of like the pandering.  I'm not sure that the release note will reach people using these, especially -XX:+TraceExceptions.  I guess they've been getting a deprecation message since 9 so maybe it won't be such a surprise.  I never stand in the way of removing code that people won't need, if you think this is the case.

Reply:
I think we should remove them. There is a risk that the user might not notice the difference between the old and new messages and so not realize they don't actually have active logging any more (until they go to access the logs to diagnose a problem).